src/disk_cache.rs

```diff
@@ -0,0 +1,956 @@
+        self.object_store.copy(from, to).await
+    }
+
+    async fn rename(&self, from: &Path, to: &Path) -> object_store::Result<()> {
+        self.object_store.rename(from, to).await
+    }
+
+    async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> object_store::Result<()> {
+        self.object_store.copy_if_not_exists(from, to).await
+    }
+
+    async fn rename_if_not_exists(&self, from: &Path, to: &Path) -> object_store::Result<()> {
+        self.object_store.rename_if_not_exists(from, to).await
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct DiskCacheMeta {
+    pub location: String,
+    pub last_modified: String,
+    pub size: usize,
+    pub e_tag: Option<String>,
+    pub version: Option<String>,
+}
+
+// it seems that object_store did not expose this error type, duplicate it here.
+// TODO: raise a pr to expose this error type in object_store.
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum InvalidGetRange {
+    #[error("Range start too large, requested: {requested}, length: {length}")]
+    StartTooLarge { requested: usize, length: usize },
+
+    #[error("Range started at {start} and ended at {end}")]
+    Inconsistent { start: usize, end: usize },
+}
+
+struct DiskCacheEntry {
+    root_folder: std::path::PathBuf,
+    location: object_store::path::Path,
+    part_size: usize,
+}
+
+#[allow(unused)]
+type PartID = usize;
+
+#[allow(unused)]
+impl DiskCacheEntry {
+    /// save the GetResult to the disk cache, a GetResult may be transformed into multiple parts
+    /// file and a meta file. please note that the `range` in the GetResult is expected to be
+    /// aligned with the part size.
+    pub async fn save_result(&self, result: GetResult) -> object_store::Result<usize> {
+        assert!(result.range.start % self.part_size == 0);
+
+        self.save_meta(&result.meta).await?;
+
+        let mut buffer = BytesMut::new();
+        let mut part_number = result.range.start / self.part_size;
+        let object_size = result.meta.size;
+
+        let mut stream = result.into_stream();
+        while let Some(chunk) = stream.next().await {
+            let chunk = chunk?;
+            buffer.extend_from_slice(&chunk);
+
+            while buffer.len() >= self.part_size {
+                let to_write = buffer.split_to(self.part_size);
+                self.save_part(part_number, to_write.as_ref()).await?;
+                part_number += 1;
+            }
+        }
+
+        // if the last part is not fully filled, save it as the last part. This is useful
+        // to determined the end of the object.
+        if !buffer.is_empty() {
+            self.save_part(part_number, buffer.as_ref()).await?;
+            return Ok(object_size);
+        }
+
+        // if reached exactly the end of the object file, save an empty part file to indicate
+        // the end of the object.
+        if part_number * self.part_size == object_size {
```

Comment:
Curious about this. Is it used anywhere? Why's it needed?

Reply:
this can be removed, it can be useful to determine the end of object if meta is not cached, but in most of the time the meta file is cached together, i think we can remove this to make it simpler.